### PR TITLE
Denormalize entity subtypes

### LINKED ISSUE #10476: Remove support for subtype-less entities
We have created so much boilerplate around normalizing user and group subtypes. How about we require subtype to be set on all entities. Upgrade user to user:user and admins to user:admin. Group to group:group.

### PR DESCRIPTION
**Pull request by hypeJunction**
Entity subtypes have been denormalized. 

Elgg no longer maintains subtype IDs, instead uses arbitrary strings.  

BREAKING CHANGE 
entity_subtypes database table has been removed.  

Entity subtypes no longer have IDs, subsequently all API around entity subtypes has been removed, including add_subtype(), update_subtype(), remove_subtype(), get_subtype_from_id(), get_subtype_id(), get_subtype_class(), get_subtype_class_from_id().  

Plugins can define custom entity classes using elgg_set_entity_class() at runtime.

Fixes #10476

- [ ] Refuse schema migration if subtypes table holds `user`, `group` or `site` subtype
- [ ] Check for objects without subtype
- [ ] Document query conventions: if `type` and `subtype` are set, the will also be search for literally (`true`, `false`, `0` will be case to string in in `e.subtype IN ()` query); if not set, all entities of type and subtype will be returned
- [ ] Needs #11055
- [ ] `register_user` should accept subtype argument

### CODE REVIEW

File: engine/classes/Elgg/Database/EntityTable.php [hunk 1]
```diff
@@ -1221,7 +1196,13 @@ public function getEntityAttributeWhereSql(array $options = []) {
 		}
 
 		// @todo the types should be defined somewhere (as constant on \ElggEntity?)
-		if (!in_array($type, ['group', 'object', 'site', 'user'])) {
+		if (!in_array($type, [
```
Comment (mrclay) [inline engine/classes/Elgg/Database/EntityTable.php]: If our CS rules think this is better I have words for it. Maybe make a $valid_types array above and use it?

Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: I think we should remove this from config and make it a static var somewhere: ``Application::$valid_types``

File: engine/classes/Elgg/AttributeLoader.php [hunk 2]
```diff
@@ -234,7 +234,7 @@ public function getRequiredAttributes(\stdClass $row) {
 
 		$row = $this->filterAddedColumns($row);
 
-		$row['subtype'] = (int) $row['subtype'];
+		$row['subtype'] = $row['subtype'];
```
Comment (hypeJunction) [inline engine/classes/Elgg/AttributeLoader.php]: Redundant

File: engine/classes/Elgg/Database/EntityTable.php (hunk 1 above)
Comment (mrclay) [inline engine/classes/Elgg/Database/EntityTable.php]: Something like that is in my branch. No worries for now.

File: engine/classes/Elgg/Database/EntityTable.php [hunk 3]
```diff
@@ -1101,7 +1072,7 @@ public function getGuidBasedWhereSql($column, $guids) {
 	 * @access private
 	 */
 	public function getEntityTimeWhereSql($table, $time_created_upper = null,
-	$time_created_lower = null, $time_updated_upper = null, $time_updated_lower = null) {
+										  $time_created_lower = null, $time_updated_upper = null, $time_updated_lower = null) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: PHPStorm being weird

**Comment by mrclay** [comment]
This is heroic. My only concerns are around the many fuzzy edge cases in our query APIs, like if you search for `type='blog', subtype=''` or `type='blog', subtype=null` do these bring back all blogs, or only those made by the blog plugin, etc. And would these now fail?

File: engine/classes/Elgg/Database/EntityTable.php (hunk 1 above)
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: This is actually PHPStorm formatting, I am not sure how to tell it to keep it as is. There were many changes in this file. I can rever some of these.

File: engine/classes/Elgg/Database/EntityTable.php [hunk 4]
```diff
@@ -1334,17 +1315,10 @@ public function getDates($type = '', $subtype = '', $container_guid = 0, $order_
...
-								return false;
-							}
-						} else {
-							$subtypeval = 0;
-						}
 						if (!empty($tempwhere)) {
 							$tempwhere .= " or ";
 						}
-						$tempwhere .= "(type = '{$typekey}' and subtype = {$subtypeval})";
+						$tempwhere .= "(type = '{$typekey}' and subtype = '{$subtypeval}'')";
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: doesn't look right

File: engine/classes/Elgg/Database/EntityTable.php [hunk 5]
```diff
@@ -1353,11 +1327,7 @@ public function getDates($type = '', $subtype = '', $container_guid = 0, $order_
 			}
 		} else {
 			if ($subtype) {
-				if (!$subtype_id = get_subtype_id($type, $subtype)) {
-					return false;
-				} else {
-					$where[] = "subtype=$subtype_id";
-				}
+				$where[] = "subtype='$subtype''";
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: extraneous `'`

File: engine/classes/ElggPlugin.php [hunk 6]
```diff
@@ -1121,7 +1117,7 @@ protected function deactivateEntities() {
 		
 		foreach ($spec as $entity) {
 			if (isset($entity['type'], $entity['subtype'], $entity['class'])) {
-				update_subtype($entity['type'], $entity['subtype']);
+				elgg_set_entity_class($entity['type'], $entity['subtype']);
```
Comment (hypeJunction) [inline engine/classes/ElggPlugin.php]: This will have no effect. Can be removed

File: engine/tests/ElggCoreGetEntitiesFromAnnotationsTest.php [hunk 7]
```diff
@@ -287,7 +288,6 @@ public function testElggGetEntitiesFromAnnotationsCalculateX() {
 						break;
 				}
 
-				$this->assertEqual($e->guid, $order[$i]);
```
Comment (hypeJunction) [inline engine/tests/ElggCoreGetEntitiesFromAnnotationsTest.php]: Not sure what this is supposed to test, but it did break the tests. Seems like order has changed somehow.

File: engine/tests/ElggEntityTest.php [hunk 8]
```diff
@@ -176,7 +161,7 @@ public function testElggEntityGetAndSetAnnotations() {
 		// @todo belongs in Annotations API test class
 		$this->assertIdentical($annotations, elgg_get_annotations(array('guid' => $this->entity->getGUID(), 'annotation_name' => 'non_existent')));
 		$this->assertIdentical($annotations, elgg_get_annotations(array('guid' => $this->entity->getGUID(), 'annotation_name' => 'non_existent', 'type' => 'object')));
-		$this->assertIdentical(false, elgg_get_annotations(array('guid' => $this->entity->getGUID(), 'type' => 'object', 'subtype' => 'fail')));
+		$this->assertIdentical([], elgg_get_annotations(array('guid' => $this->entity->getGUID(), 'type' => 'object', 'subtype' => 'fail')));
```
Comment (hypeJunction) [inline engine/tests/ElggEntityTest.php]: `elgg_get_annotations()` returns an array. Not sure how this was passing

**Comment by hypeJunction** [comment]
> This is heroic. My only concerns are around the many fuzzy edge cases in our query APIs, like if you search for type='group', subtype='' or type='group', subtype=null do these bring back all groups, or only those made by the group plugin, etc. And would these now fail? Same with elgg_instanceof($group, 'group', '');

All falsey subtype values will fail to save an entity. I left the ege* queries relaxed, so if you search for `subtype=0` it will search it. AFAICS, our API was returning all entities, if you were searching for `type='invalid'`, this actually brings back some sanity to SQL.

**Comment by hypeJunction** [comment]
@mrclay All tests have passed locally. Any idea what's happening on Travis, namely why `Config` is sending headers?

**Comment by jdalsem** [comment]
> Plugins can define custom entity classes using elgg_set_entity_class() at runtime.

Don't introduce this... can do this already via elgg-plugin.php -> http://learn.elgg.org/en/master/guides/database.html?highlight=elgg-plugin.php#extending-elggentity

**Comment by hypeJunction** [comment]
> Don't introduce this... can do this already via elgg-plugin.php

What if you want to replace another plugin's class definition?

**Comment by jdalsem** [comment]
> What if you want to replace another plugin's class definition?

just include it in your own elgg-plugin.php... last one wins

**Comment by hypeJunction** [comment]
Don't like. I want declarative API that can be run with whatever priority I please

**Comment by jdalsem** [comment]
> Don't like. I want declarative API that can be run with whatever priority I please

on second thought i agree... could you add a bit of docs for this function (maybe after the docs section on entity registration in elgg-plugin.php)

**Comment by jdalsem** [comment]
there is a typo in your 'denormalize' commit text

File: engine/tests/ElggCoreGetEntitiesFromAnnotationsTest.php [hunk 9]
```diff
@@ -215,6 +215,7 @@ public function testElggGetEntitiesFromAnnotationsCalculateX() {
 				'type' => 'object',
 				'subtypes' => $subtypes,
 				'limit' => $num_entities,
+				'order_by' => 'e.time_created ASC',
```
Comment (jdalsem) [inline engine/tests/ElggCoreGetEntitiesFromAnnotationsTest.php]: why was this needed?

File: engine/tests/ElggCoreGetEntitiesFromAnnotationsTest.php (hunk 7 above)
Comment (jdalsem) [inline engine/tests/ElggCoreGetEntitiesFromAnnotationsTest.php]: maybe your custom order by changed the order

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php [hunk 10]
```diff
@@ -638,7 +638,7 @@ public function testCanOverrideContainerLogicWithAHook() {
 	 */
 	public function testCanDownloadFileByDefault() {
 
-		add_subtype('object', 'file', \ElggFile::class);
+		elgg_set_entity_class('object', 'file', \ElggFile::class);
```
Comment (jdalsem) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: is this needed? core sets the class already am i right?

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php [hunk 11]
```diff
@@ -658,7 +658,7 @@ public function testCanDownloadFileByDefault() {
 	 */
 	public function testCanAlterDownloadPermissions() {
 
-		add_subtype('object', 'file', \ElggFile::class);
+		elgg_set_entity_class('object', 'file', \ElggFile::class);
```
Comment (jdalsem) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: same here

File: engine/tests/suite.php [hunk 12]
```diff
@@ -66,7 +109,7 @@
 }
 
 // disable emails
-elgg_set_email_transport('mailer', new Zend\Mail\Transport\InMemory());
+elgg_set_email_transport(new Zend\Mail\Transport\InMemory());
```
Comment (jdalsem) [inline engine/tests/suite.php]: was this bugged or something?

File: mod/file/elgg-plugin.php [hunk 13]
```diff
@@ -5,6 +5,7 @@
 		[
 			'type' => 'object',
 			'subtype' => 'file',
+			'class' => '\ElggFile',
```
Comment (jdalsem) [inline mod/file/elgg-plugin.php]: is this needed because it will be overriden otherwise?

**Comment by jdalsem** [comment]
This looks very nice... happy to see this happening

File: mod/file/elgg-plugin.php (hunk 13 above)
Comment (hypeJunction) [inline mod/file/elgg-plugin.php]: Not needed . As long as class is not set, it won't be replaced.

File: engine/tests/suite.php (hunk 12 above)
Comment (hypeJunction) [inline engine/tests/suite.php]: Yes, I don't know why it's not short-circuting on Travis

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php (hunk 10 above)
Comment (hypeJunction) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: We don't run init, system in phpunit

File: engine/tests/ElggCoreGetEntitiesFromAnnotationsTest.php (hunk 7 above)
Comment (hypeJunction) [inline engine/tests/ElggCoreGetEntitiesFromAnnotationsTest.php]: No, they were failing even before I added order clauses

File: engine/schema/migrations/20170804152152_align_subtype_columns.php [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+		$options = [
+			'null' => false,
+			'limit' => 252,
+			'encoding' => "utf8",
+			'collation' => "utf8_general_ci",
+		];
+
+		$table = $this->table('river');
+		$table->changeColumn('subtype', 'string', $options)->save();
```
Comment (Wouter0100) [inline engine/schema/migrations/20170804152152_align_subtype_columns.php]: As seen in the docblock above, the `change()` method does not support `changeColumn()`, as it's unable to track back the changes for migrating down. I would suggest using `up()` and `down()` instead in case `change()` does not support it to fully support rolling back.

**Comment by Wouter0100** [comment]
Happy to see such changes! Thank you guys. Can't wait for the next big release.

**Comment by mrclay** [comment]
Rebase hell! But in the final stretch.

File: engine/schema/migrations/20170804152152_align_subtype_columns.php (hunk 14 above)
Comment (hypeJunction) [inline engine/schema/migrations/20170804152152_align_subtype_columns.php]: These commits were merged in core already. Feel free to break down

**Comment by hypeJunction** [comment]
I think you need to wipe the phinx commits before rebasing

**Comment by hypeJunction** [comment]
Rebased now
